File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
const par = [];
const sz = [];

function find(v){
    if(par[v] == v) return v;
    else return par[v] = find(par[v]);
}

function union(a, b){
    a = find(a);
    b = find(b);
    
    if(a == b) return;
    sz[b] += sz[a];
    par[a] = b;
}

function solution(n, wires) {
    let res = 1e9;
    for(let j = 0; j < wires.length; j++){
        for(let i = 1; i <= n; i++) {
            par[i] = i;
            sz[i] = 1;
        }
        
        for(let i = 0; i < wires.length; i++){
            if(i == j) continue;
            const [a,b] = wires[i];

            union(a,b);
        }
        const [a,b] = wires[j];
        const diff = Math.abs(sz[find(a)] - sz[find(b)]);
        res = Math.min(diff, res);
    }
    
    return res;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

const int MX = 102;
vector<int> vec[MX];

int a, b;
int par[MX];
int sz[MX];

int find(int a){
    if(par[a] == a) return a;
    else return par[a] = find(par[a]);
}

void union_(int a, int b){
    a = find(a);
    b = find(b);
    if(a==b) return;
    sz[b] += sz[a];
    par[a] = b;
}

int solution(int n, vector<vector<int>> wires) {
    int mn = 1e9;
    
    for(int i = 0; i < wires.size(); i++){
        for(int j = 0; j < MX; j++){
            par[j] = j;
            sz[j] = 1;
        }
        int fa, fb;
        for(int j = 0; j < wires.size(); j++){
            int a = wires[j][0];
            int b = wires[j][1];
            if(j == i){
                fa = a;
                fb = b;
                continue;
            } else {
                union_(a,b);
            } 
        }
        mn = min(mn, abs(sz[find(fa)] - sz[find(fb)]));
    }
    
    return mn;
}